File: c/Caesar.c
/*
 * Caesar Cipher
 *
 * @author  lellansin <lellansin@gmail.com>
 * @website http://www.lellansin.com/tutorials/ciphers
 */

#include <stdio.h>
#include <ctype.h>

char getShiftCh(char ch, int shift, char start, char end);

/*
 * 凯撒密码
 *
 * @param char * [in]  words  要加密的字符串
 * @param int    [in]  shift  位移
 * @param char * [out] result 结果保存
 */
void caesar(char *words, int shift, char *result)
{
    int i;
    char *p = words;

    for (i = 0; *p; p++, i++)
    {
        if (isalpha(words[i]))
        {
            if (islower(words[i]))
            {
                sprintf(result, "%s%c", result, getShiftCh(words[i], shift, 'a', 'z'));
            } else
            {
                sprintf(result, "%s%c", result, getShiftCh(words[i], shift, 'A', 'Z'));
            }
        } else
        {
            sprintf(result, "%s%c", result, words[i]);
        }
    }
}

char getShiftCh(char ch, int shift, char start, char end)
{
    int diff = ((ch - start) + shift) % 26;
    if (diff >= 0) {
        return  start + diff;
    }
    else {
        return end + diff + 1;
    }
}

int main(int argc, char const *argv[])
{
    char text[] = "hello, world!";
    char cipherstext[1024], result[1024];
    int shift = -5;

    // 左移 5
    caesar(text, shift, cipherstext);
    printf("左移 %s\n", cipherstext);

    // 右移 5
    caesar(cipherstext, -shift, result);
    printf("右移 %s\n", result);

    return 0;
}

File: erlang/caesar.erl
%
% Caesar Cipher
% 
% @author  lellansin <lellansin@gmail.com>
% @website http://www.lellansin.com/tutorials/ciphers
%
-module(caesar).
-export([encrypt/2, decrypt/2, test/0]).

% 
% 加密
% 
encrypt(Words, Shift) -> [ shift(C, Shift) || C <- slashes(Words) ].

% 
% 解密
% 
decrypt(Words, Shift) -> encrypt(Words, -Shift).

% 
% 转义
% 
slashes(Words) ->
    re:replace(Words, "~n", "\n", [ global, { return, list } ]).

% 
% 位移
% 
shift(C, Shift) when ((C >= $A) and (C =< $Z)) ->
    Offset = Shift rem 26,
    if
        Offset + C > $Z  -> $A + (Offset + C) - $Z;
        Offset + C < $A  -> $Z + Offset + C;
        true -> C + Offset
    end;
shift(C, Shift) when ((C >= $a) and (C =< $z)) ->
    Offset = Shift rem 26,
	if
        Offset + C > $z  -> $a + (Offset + C) - $z;
        Offset + C < $a  -> $z + Offset + C;
        true -> C + Offset
    end;
shift(C, _) -> C.

% 
% 测试
% 
test() ->
    Text = "hello world, this is Caesar cipher.~n",
    Cipher = encrypt(Text, 3),
    io:format(Cipher),
    io:format(decrypt(Cipher, 3)).
